pkg/finality-grandpa/voter.go

```diff
@@ -0,0 +1,1125 @@
+					}
+				}
+
+				err := v.env.Completed(
+					justCompleted.roundNumber(),
+					justCompleted.roundState(),
+					justCompleted.dagBase(),
+					justCompleted.historicalVotes(),
+				)
+				if err != nil {
+					v.inner.Unlock()
+					return err
+				}
+
+				v.inner.pastRounds.Push(v.env, justCompleted)
+
+				oldBest := v.inner.bestRound
+				v.inner.bestRound = newBest
+				v.inner.pastRounds.Push(v.env, oldBest)
+
+				if processCatchUpOutcome != nil {
+					processCatchUpOutcome(newCatchUpProcessingOutcome(CatchUpProcessingOutcomeGood{}))
+				}
+				v.inner.Unlock()
+			}
+		default:
+			break loop
+		}
+	}
+	return nil
+}
+
+// process the logic of the best round.
+func (v *Voter[Hash, Number, Signature, ID]) processBestRound(waker *waker) (bool, error) {
+	// If the current `bestRound` is completable and we've already precommitted,
+	// we start a new round at `bestRound + 1`.
+	{
+		v.inner.Lock()
+
+		var shouldStartNext bool
+		completable, err := v.inner.bestRound.poll(waker)
+		if err != nil {
+			return true, err
+		}
+
+		var precomitted bool
+		state := v.inner.bestRound.State()
+		if state != nil {
+			_, precomitted = v.inner.bestRound.State().(statePrecommitted)
+		}
+
+		shouldStartNext = completable && precomitted
+
+		if !shouldStartNext {
+			v.inner.Unlock()
+			return false, nil
+		}
+
+		log.Tracef("Best round at %v has become completable. Starting new best round at %v",
+			v.inner.bestRound.roundNumber(),
+			v.inner.bestRound.roundNumber()+1,
+		)
+		v.inner.Unlock()
+	}
+
+	err := v.completedBestRound()
+	if err != nil {
+		return true, err
+	}
+
+	// round has been updated. so we need to re-poll.
+	return v.poll(waker)
+}
+
+func (v *Voter[Hash, Number, Signature, ID]) completedBestRound() error {
+	v.inner.Lock()
+	defer v.inner.Unlock()
+
+	err := v.env.Completed(
+		v.inner.bestRound.roundNumber(),
+		v.inner.bestRound.roundState(),
+		v.inner.bestRound.dagBase(),
+		v.inner.bestRound.historicalVotes(),
+	)
+	if err != nil {
+		return err
+	}
+
+	oldRoundNumber := v.inner.bestRound.roundNumber()
+
+	nextRound := newVotingRound(
+		oldRoundNumber+1,
+		v.voters,
+		v.lastFinalizedInRounds,
+		v.inner.bestRound.bridgeState(),
+		v.inner.bestRound.FinalizedSender(),
+		v.env,
+	)
+
+	oldBest := v.inner.bestRound
+	v.inner.bestRound = nextRound
+	v.inner.pastRounds.Push(v.env, oldBest)
+	return nil
+}
+
+func (v *Voter[Hash, Number, Signature, ID]) setLastFinalizedNumber(finalizedNumber Number) bool {
+	if finalizedNumber > v.lastFinalizedNumber {
+		v.lastFinalizedNumber = finalizedNumber
+		return true
+	}
+	return false
+}
+
+func (v *Voter[Hash, Number, Signature, ID]) Start() error { //skipcq: RVV-B0001
+	v.wg.Add(1)
+	defer v.wg.Done()
+	waker := newWaker()
+	for {
+		ready, err := v.poll(waker)
+		if err != nil {
+			return err
+		}
+		if ready {
+			return nil
+		}
+		select {
+		case <-waker.channel():
+		case <-v.stopChan:
+			return fmt.Errorf("early voter stop")
```

Comment:
sentinel error

Reply:
so I can check with `errors.Is/As`?

Reply:
I don't think this needs to be a public sentinel error given that the caller will most likely not actually inspect this error and make any different decisions.  For the case of the `Voter.Start()` returning this specific error, this will almost always be intentionally stopped early when the node shuts down and so this error will be suppressed.  This specific error may not actually need to be returned in hindsight.  All other errors returned by `Voter.Start()` would be catastrophic errors which would trigger a graceful shutdown and logged out as a critical error.  If you're suggesting make this a private sentinel error, I tried to articulate when I would do so in this [thread](https://github.com/ChainSafe/gossamer/pull/4126#discussion_r1732143634).